### PR TITLE
Couple training budgets, schedules, and validation stopping

### LINKED ISSUE #67: Couple lr schedule and stopping rule into fixed_budget / open_ended modes
## Summary

`optimize()` exposes the learning-rate schedule and the stopping rule as independent
settings, and the default combination of the two is not coherent:

- `epochs=200` (default)
- `early_stopping_patience=5` (default)
- `decay_steps = num_batches * epochs` — cosine annealing to zero over the full 200
  epochs (`optimization/objective.py:667`)

A trial that early-stops around epoch 20 has annealed roughly 1% of that horizon, so it
stops at essentially peak learning rate. The returned model is not "the 20-epoch model";
it is a model frozen mid-flight, worse than a run actually scheduled for 20 epochs would
have been. Because early stopping *keeps* the model it stops — unlike pruning, which
discards it — this degrades the trial result rather than merely wasting compute.

Proposal: replace the two independent knobs with one **mode**, so schedule and stopping
are always chosen as a coherent pair, and fix three correctness issues that fall out.

## Proposed modes

|              | `fixed_budget`                            | `open_ended`                           |
| ------------ | ----------------------------------------- | -------------------------------------- |
| lr schedule  | cosine to zero                            | inverse-sqrt with warmup               |
| horizon      | the trial's actual `epochs * num_batches` | none                                   |
| stopping     | runs to budget                            | early stop on the validation objective |
| `epochs`     | derived from a simulation budget          | a generous cap, not a target           |
| pruning      | allowed (discards, so safe)               | allowed                                |

**Invariant to enforce:** annealing to a horizon and stopping before that horizon are
mutually exclusive. Setting `early_stopping_patience` together with a horizon-annealed
schedule should warn or raise.

**Suggested default: `fixed_budget`.** Under early stopping trials receive different
budgets, so "best trial" conflates configuration quality with how much training that
trial happened to get. Equal-budget trials are directly comparable; early-stopped ones
are not. `open_ended` is the right choice when the output is "the best model I can get"
and per-configuration convergence differences are large or unknown; `fixed_budget` is
right when the output is a *ranking* that has to be defended.

## Composing the two modes

The modes are most useful in sequence rather than as an either/or, and the API should
make that cheap (for example a mode switch on an otherwise resumable study):

1. **Screen — `open_ended`.** Wide space, one seed per trial, early stopping. Cheap per
   trial and tolerant of configurations whose convergence speed differs a lot. Its
   output is a *shortlist*, not a ranking — unequal budgets make the ordering unsafe to
   report.
2. **Confirm — `fixed_budget`.** Re-run the top ~5 at equal budget with several seeds,
   under the schedule the final model will actually use. Rank here.

This also addresses a separate problem: within-configuration seed noise in our study is
sd ~0.005-0.009 while the spread across the best architectures is ~0.010, so a
single-seed leaderboard is substantially seed lottery. Putting the replication only in
stage 2 costs roughly `N + 15` runs instead of `3N`, and puts it where the decision is.

It also removes the main objection to `open_ended` ranking under a different schedule
than the final fit: stage 2 re-fits under the production schedule anyway.

### 1. Schedule horizon can diverge from actual training length

```python
# optimization/objective.py:667
decay_steps = config.num_batches * config.epochs   # study-level constants
```

but a custom `train_fn` trains for the per-trial values. In `bayesflow-irt`:

```python
# bayesflow_irt/hpo.py:536-538
epochs=int(hparams["epochs"]),
num_batches=int(hparams["num_batches"]),
```

While these agree the mismatch is invisible. As soon as training length varies per trial
— which `fixed_budget` requires — every trial trains under a schedule annealed for the
wrong length. Fix: derive `decay_steps` from the same values `train_fn` uses.

### 2. Early stopping monitors training loss

`EarlyStoppingOnMA(monitor="loss")` stops on a moving average of the **training** loss.
Measured counter-example from the `bayesflow-irt` 2PL study (30x500 corner, recovery
correlation for `log(a)`):

| configuration               | final training loss | recovery          |
| --------------------------- | ------------------- | ----------------- |
| `v2_irc_sabs2_lr1e3`        | 0.5705              | 0.930             |
| `v2_sabs0_h64b4_bs32_dur2x` | **0.5622** (better) | **0.912** (worse) |

Training loss inverts against the objective across architectures — the axis HPO varies.
Pruning is unaffected: `PeriodicValidationCallback` already reports the validation
objective via `trial.report()`. Early stopping does not, and it keeps the model it
stops. Fix: default early stopping to that same validation objective.

### 3. No derived hyperparameters

In online SBI `batch_size` is simultaneously the data-volume knob:
`simulations = batch_size * steps`. At most two of {batch, steps, simulations} can be
fixed. The meaningful choice is to fix the **simulation budget** — it is proportional to
GPU-seconds, which is what a run costs — and let steps follow:

```
num_batches = sims_budget // (batch_size * epochs)
```

Today that arithmetic can only live in a custom `train_fn`, where it collides with issue
1, since the scheduler would not know about it. A first-class derived dimension would let
the budget be declared and have steps and epochs fall out.

Motivation: tuning `epochs` and `lr` as independent dimensions is close to redundant.
Under cosine annealing the optimizer's total displacement scales roughly with
`lr_peak * steps`, so the two form a ridge — and TPE models parameters marginally, which
is the worst case for a ridge. Measured: `lr=2e-3` at 16 epochs matched `lr=5e-4` at 32
epochs, and 64 epochs added nothing further. Making `epochs` derived removes the ridge
rather than asking the sampler to navigate it.

## Explicitly not requested

Conditional or nested search spaces (an `lr` range depending on the sampled
`batch_size`). The coupling has a known functional form — Smith et al. (2018) show the
optimum scales as `B` proportional to `epsilon` — so the right treatment is
reparametrisation, `lr = lr_ref * (batch_size / reference_batch)`, keeping the space
rectangular. A rectangle in decoupled coordinates is easier for TPE than a conditional
space, not merely easier to implement.

## References

- Smith, S. L., Kindermans, P.-J., Ying, C., & Le, Q. V. (2018). *Don't decay the
  learning rate, increase the batch size.* ICLR.
  https://doi.org/10.48550/arXiv.1711.00489
- Shallue, C. J., Lee, J., Antognini, J. M., Sohl-Dickstein, J., Frostig, R., & Dahl,
  G. E. (2019). Measuring the effects of data parallelism on neural network training.
  *JMLR, 20*(112), 1-49. https://doi.org/10.48550/arXiv.1811.03600 — Sec. 2: a learning
  rate tuned at one batch size and reused at a larger one produces either divergence or
  failure to reach the error goal within a fixed epoch budget.
- Draxler, F., Sorrenson, P., Zimmermann, L., Rousselot, A., & Köthe, U. (2023).
  *Free-form flows.* App. C.1, Tables 3-4 — grid-searches batch size (2^2 to 2^8) jointly
  with learning rate, per dataset *and per simulation budget*, on the SBI benchmark.
  Selected batch grows with budget (median 8 / 32 / 64 across the three budgets), which
  is why batch size should be tunable rather than a constant.

## Status

The 3x3 batch x learning-rate grid this issue was filed ahead of has **completed**
(9 arms x 3 seeds, simulation-matched at 102,400 simulations). Results are in
https://github.com/matthiaskloft/bayesflow-hpo/issues/67#issuecomment-5349211054.

Headline changes to the proposal above, from that data:

- The objective **cannot be recovery correlation alone** — the worst-calibrated cell in the
  grid ranks 4th of 9 on recovery. This reinforces correctness issue 2.
- `batch_size` should not stay `IntDimension(constant=256)`: the batch x lr interaction is
  invisible at constant batch, and it is the largest effect measured.
- The `initial_lr` upper bound of 5e-3 is likely too low, though the higher rate trades
  `b` calibration against `a` calibration at few items rather than dominating.

— Claude Code (Opus 5)

### PR DESCRIPTION
**Pull request by matthiaskloft**
## Summary

- introduce coherent `fixed_budget` and `open_ended` training modes with matching learning-rate schedules and stopping behavior
- derive schedule horizons from each trial's actual training budget and add derived search-space dimensions
- add source-backed warmup controls, including a fixed 5% cosine warmup default with a 10% ceiling
- tune batch size over 32–256 and widen peak learning-rate exploration to `1e-2`
- stop open-ended training on `early_stopping_monitor="objective_mean"` and restore the best validation weights
- address independent-review findings for higher-is-better contraction, duplicate dimension names, and API documentation

Closes #67.

## Test plan

- `KERAS_BACKEND=torch python -m pytest tests/ -v --tb=short` — 751 passed
- `ruff check src/ tests/` — clean

— Codex (GPT-5)

## Summary by CodeRabbit

* **New Features**
  * Added fixed-budget and open-ended training modes.
  * Added configurable early stopping, validation monitoring, best-weight restoration, and learning-rate warmup.
  * Added derived search-space dimensions and expanded batch-size and learning-rate tuning ranges.
  * Added inverse-square-root learning-rate scheduling for open-ended training.

* **Bug Fixes**
  * Improved validation of warmup settings and duplicate search-space parameters.
  * Standardized higher-is-better contraction metrics for optimization.

* **Documentation**
  * Updated API, configuration, architecture, optimization, and reference documentation.

### CODE REVIEW

**Comment by matthiaskloft** [comment]
## Review: training budget modes

Reviewed the `src/` diff at `7a207ca`. The core schedule work holds up — I verified the piece that's easiest to get wrong: Keras `CosineDecay` treats `decay_steps` as the **post-warmup** horizon (total = `decay_steps + warmup_steps`, per `keras/src/optimizers/schedules/learning_rate_schedule.py`), so subtracting `warmup_steps` in `_make_cosine_decay_optimizer` is correct and the schedule really does land at ~0 exactly at `epochs * num_batches`. The `initial_learning_rate=0.0` + `warmup_target` construction is also right given `alpha=0.0`. `InverseSqrtDecay` matches Vaswani et al. §5.3 and round-trips through `get_config`/`from_config`.

Findings below, most actionable first.

### 1. `HIGHER_IS_BETTER` fixes `contraction` but the same bug remains for `coverage_*`

`_metric_to_minimize` silently treats *any* metric not in `HIGHER_IS_BETTER` as minimize-is-better. `coverage` is in `DEFAULT_METRICS` and emits `coverage_90`/`coverage_95`/`coverage_975`/`coverage_99` (`validation/registry.py:503-573`). A user who puts `coverage_95` in `objective_metrics` gets it minimized — the optimizer is rewarded for driving empirical coverage toward 0. And coverage isn't fixable by adding it to `HIGHER_IS_BETTER` either: the right objective is `|empirical − nominal|`, which is neither direction.

Now that this direction metadata is load-bearing for `early_stopping_monitor="objective_mean"` (an unknown-direction metric silently corrupts the stopping signal too, not just the final objective), I'd suggest attaching direction to the metric registry alongside `_DESCRIPTIONS`/`_KINDS` — `"minimize" | "maximize" | "target"` — and having `ObjectiveConfig.__post_init__` reject an `objective_metrics` entry with no declared direction rather than defaulting to minimize.

Separately: flipping `contraction` changes objective values for any existing study database. Worth a line in the PR description or `docs/optimization.md` that old studies aren't comparable across this change.

### 2. `early_stopping_window` now means two different things

The same config value feeds `MovingAverageEarlyStopping` (window in **epochs**) and `PeriodicValidationCallback` (window in **validation checks**). With the defaults — `early_stopping_window=7`, `intermediate_validation_interval=10`, `intermediate_validation_warmup=10`, `epochs=200` — open-ended training gets 20 validation checks total, an MA smoothing over 70 epochs, and `patience=5` costing another 50. Earliest possible stop is ~epoch 60, and realistically the criterion rarely fires before the 200-epoch cap, so `open_ended` mostly degenerates into "fixed budget with inverse-sqrt decay."

Either give the validation stopper its own knob (`early_stopping_validation_window`, default ~3), or keep one knob expressed in epochs and convert via `interval`. The API docstring ("Moving-average window size for early stopping (default 7)") should say which unit applies in which mode either way.

### 3. Duplicate-dimension detection stops at the space boundary

`BaseSearchSpace.dimensions` now catches duplicates *within* one space — good — but `CompositeSearchSpace.sample` merges three spaces with plain `dict.update` (`search_spaces/composite.py:72-77`), so a name defined in both e.g. `TrainingSpace` and a user inference space silently overwrites. That's more likely now that `DerivedDimension` invites users to define `epochs`/`num_batches` in their own space, and if both sides call `trial.suggest_*` with the same name and different ranges Optuna will disagree with itself. Worth lifting the same check into `CompositeSearchSpace`.

### 4. Warmup-vs-budget consistency is only checked at compile time

`__post_init__` validates the warmup minimum but never that warmup fits the budget. If a user sets `lr_warmup_epochs` ≥ `epochs`, the `ValueError` from `_make_cosine_decay_optimizer` is swallowed by `_reject_compile` and **every** trial returns a penalty — a config typo that looks like a modeling failure. Since warmup steps can now vary per trial (`DerivedDimension` on `num_batches`), you can't validate all cases up front, but the static case (scalar `lr_warmup_epochs`/`lr_warmup_steps` against config `epochs`/`num_batches`) is cheap to check in `__post_init__`, and the compile-time rejection path could log at `warning` with the actual numbers.

### 5. Smaller items

- `_normalize_warmup_spec` uses `type(value) is int`, which rejects `np.int64` — plausible input from a config file or an array-derived budget. `isinstance(value, numbers.Integral) and not isinstance(value, bool)` covers the bool exclusion you're after without the false negative.
- `validation_callback.py` imports the private `_metric_to_minimize` from `objectives`. Given it's now used across modules and is central to finding #1, promoting it to public API seems right.
- `PeriodicValidationCallback` snapshots full weights via `get_weights()` on every improvement. Dropping `restore_best_weights` from `MovingAverageEarlyStopping` frees a copy so this is roughly net-neutral, but it isn't reflected in `estimate_peak_memory_mb` / the `max_memory_mb` pre-check.

Nothing here blocks the schedule/mode coupling itself — #1 is the one I'd want resolved before merge, since it's the same defect class the PR set out to fix.

— Claude (Opus 5)
